README
NAME
    with-sound - Command executor with sound

VERSION
    This document describes with-sound version 1.2.0

SYNOPSIS
        $ with-sound [command] ([argument(s)])

        # e.g.
        #    $ with-sound prove t/

DESCRIPTION
    with-sound is the command executor with sound (now with-sound can play
    the MP3 file only).

    If exit-status of the command is success, it playbacks the sound of
    success. Else, it plays sound of failure. And this application also can play the sound during executing the command.

CONFIGURATION AND ENVIRONMENT
    App::WithSound requires a resource file as '.with-soundrc' on your home
    directory. This resource file specifies the path to sound files. Like
    so:

        SUCCESS=~/Music/success.mp3
        FAILURE=~/Music/failure.mp3
        RUNNING=~/Music/running.mp3

    If you want to play the different sound corresponded to discrete command, please use section notation as the following.

        [prove]
        SUCCESS=~/Music/success_prove.mp3
        FAILURE=~/Music/failure_prove.mp3
        RUNNING=~/Music/running_prove.mp3

    In case of the above example, configured sounds will be played if you execute 'prove'.
    And if other commands are executed, this application uses default settings.
    ("Default settings" means settings that is not specified by section)

    You also can specify the paths by environment variables $ENV{'WITH_SOUND_SUCCESS'}, $ENV{'WITH_SOUND_FAILURE'}
    and $ENV{'WITH_SOUND_RUNNING'}.

    If the file paths are set by both environment variables and a
    configuration file, the environment variables are prior to the
    configuration file.

    These sound file settings are optional. You do not have to fill the all of settings.
    with-sound will ignore playing sound if the sound file is not specified.

    This module also requires "mpg123" or "mpg321" or "afplay" so please install it before use this
    application ("afplay" is pre-installed if you are using Mac OS X).

    Please refer to the following:

    http://www.mpg123.de/index.shtml

    http://mpg321.sourceforge.net/

DEPENDENCIES
    Config::Simple (version 4.58 or later)

    File::Spec::Functions (version 3.40 or later)

    File::Path::Expand (version 1.02 or later)

    File::Which (version 1.09 or later)

    Test::Warn (version 0.24 or later)

    Test::MockObject::Extends (version 1.20120301 or later)

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests to
    `bug-app-withsound@rt.cpan.org', or through the web interface at
    http://rt.cpan.org.

AUTHOR
    moznion `<moznion@gmail.com>'

    Shinpei Maruyama `<shinpeim[at]gmail.com>'

CONTRIBUTOR
    Syohei YOSHIDA `<syohex[at]gmail.com>'

    papix `<mail[at]papix.net>'

LICENCE AND COPYRIGHT
    Copyright (c) 2013, moznion `<moznion@gmail.com>'. All rights reserved.

    This module is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself. See perlartistic.

DISCLAIMER OF WARRANTY
    BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
    FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
    OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
    PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
    EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
    ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
    YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
    NECESSARY SERVICING, REPAIR, OR CORRECTION.

    IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
    WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
    REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE LIABLE
    TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR
    CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE
    SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
    RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
    FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
    SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
    DAMAGES.



Changes
Revision history for App-WithSound

{{$NEXT}}

1.2.1 2015-08-03T16:04:19Z
    - Fix a bug which is caused by hash randomization (@papix)

1.2.0  Thu Mar 14 00:42:43 JST 2013
    - Changed using environment variables name.
    - Added a function to play the different sound corresponded to discrete command.

1.1.0  Wed Mar 06 00:45:10 JST 2013
    - Add a new function that playbacks sound during executing the command.

1.0.2  Tue Mar 05 02:03:55 JST 2013
    - Corresponded plural audio playing application.

1.0.1  Fri Mar 01 08:11:20 JST 2013
    - Reworded description.

1.0.0  Fri Mar 01 01:00:52 JST 2013
    - Initial release.


lib/App/WithSound.pm
package App::WithSound;

use warnings;
use strict;
our $VERSION = '1.2.0';

use Carp;
use Config::Simple;
use File::Path::Expand;
use File::Which;
use File::Spec::Functions qw/devnull/;
use IPC::Open3;

our $SIGTERM = 15;

sub new {
    my ( $class, $config_file_path, $env ) = @_;
    bless {
        config_file_path   => $config_file_path,
        env                => $env,
        success_sound_path => undef,
        failure_sound_path => undef,
        running_sound_path => undef,
        sound_player       => undef,
    }, $class;
}

sub run {
    my ( $self, @argv ) = @_;
    unless (@argv) {
        croak 'Usage: $ with-sound [command] ([argument(s)])' . "\n";
    }

    $self->_init( $argv[0] );

    my $retval = $self->_execute_command(@argv);
    $retval = 1 if $retval > 255;

    $self->_play_sound($retval);
    return $retval;
}

sub _init {
    my ( $self, $command ) = @_;

    $self->_load_sound_paths($command);
    $self->_detect_sound_play_command;

    return $self;
}

sub _execute_command {
    my ( $self, @argv ) = @_;

    my $pid    = $self->_play_sound;
    my $retval = system(@argv);
    kill( $SIGTERM, $pid ) if $pid;

    return $retval;
}

sub _detect_sound_play_command {
    my ($self) = @_;

    my $player;
    $player ||= which('mpg123');
    $player ||= which('mpg321');
    $player ||= which('afplay');

    $self->{sound_player} = $player;
    return $self;
}

sub _load_sound_paths_from_env {
    my ($self) = @_;

    my %deprecated_envs = (
        WITH_SOUND_SUCCESS => "success_sound_path",
        WITH_SOUND_FAILURE => "failure_sound_path",
        WITH_SOUND_RUNNING => "running_sound_path",
    );
    for my $env_name ( sort keys %deprecated_envs ) {
        if ( my $sound_file_path = $self->{env}->{$env_name} ) {
            carp
qq{[WARNING] "$env_name" is deprecated. Please use "PERL_$env_name"\n};
            $self->{ $deprecated_envs{$env_name} } =
              expand_filename($sound_file_path);
        }
    }

    my %envs = (
        PERL_WITH_SOUND_SUCCESS => "success_sound_path",
        PERL_WITH_SOUND_FAILURE => "failure_sound_path",
        PERL_WITH_SOUND_RUNNING => "running_sound_path",
    );
    for my $env_name ( sort keys %envs ) {
        if ( my $sound_file_path = $self->{env}->{$env_name} ) {
            $self->{ $envs{$env_name} } = expand_filename($sound_file_path);
        }
    }

    $self;
}

sub _load_sound_paths_from_config {
    my ( $self, $command ) = @_;

    $command ||= '';

    # Not exists config file.
    unless ( -f $self->{config_file_path} ) {
        carp
          "[WARNNING] Please put config file in '$self->{config_file_path}'\n";
        return;
    }
    my %config;
    eval { Config::Simple->import_from( $self->{config_file_path}, \%config ) };
    print STDERR "Configuration file has some errors."
      . "Please check your '.withsound-rc' file.\n"
      . "(Didn't you write plural format in configuration file?)\n"
      if $@;

    $self->{success_sound_path} =
      expand_filename( $config{"$command.SUCCESS"}
          || $config{'default.SUCCESS'}
          || $config{'SUCCESS'} );
    $self->{failure_sound_path} =
      expand_filename( $config{"$command.FAILURE"}
          || $config{'default.FAILURE'}
          || $config{'FAILURE'} );
    $self->{running_sound_path} =
      expand_filename( $config{"$command.RUNNING"}
          || $config{'default.RUNNING'}
          || $config{'RUNNING'} );
    $self;
}

sub _load_sound_paths {
    my ( $self, $command ) = @_;
    $self->_load_sound_paths_from_config($command);

    # load from env after config so environment variables are prior to config
    $self->_load_sound_paths_from_env;
    $self;
}

sub _play_mp3_in_child {
    my ( $self, $play_command, $mp3_file_path ) = @_;

    my ( $devnull, $pid );
    unless ( open( $devnull, '>', devnull ) ) {
        carp "[WARNING] Couldn't open devnull : $!";
        return;
    }
    eval {
        my $wtr;
        $pid = open3( $wtr, '>&' . fileno($devnull),
            0, $play_command, $mp3_file_path, );
        close $wtr;
    };
    carp "[WARNING] Couldn't exec $play_command in sound process: $@" if $@;
    return $pid;
}

sub _play_mp3 {
    my ( $self, $mp3_file_path, $status ) = @_;

    return unless $mp3_file_path;

    # not exists mp3 file
    unless ( -f $mp3_file_path ) {
        carp "[WARNING] Sound file not found for $status. : $mp3_file_path";
        return;
    }

    my $play_command = $self->{sound_player};
    unless ($play_command) {
        carp "[WARNING] No sound player is installed."
          . "please install mpg123 or mpg321";
        return;
    }

    $self->_play_mp3_in_child( $play_command, $mp3_file_path );
}

sub _play_sound {
    my ( $self, $command_retval ) = @_;

    my $pid;
    if ( !defined($command_retval) ) {

        # running
        $pid = $self->_play_mp3( $self->{running_sound_path}, 'running' );
    }
    elsif ( $command_retval == 0 ) {

        # success
        $pid = $self->_play_mp3( $self->{success_sound_path}, 'success' );
    }
    else {
        # failure
        $pid = $self->_play_mp3( $self->{failure_sound_path}, 'failure' );
    }
    return $pid;
}

1;
__END__

=encoding utf8

=head1 NAME

App::WithSound - Execute commands with sound


=head1 VERSION

This document describes App::WithSound version 1.2.0


=head1 DESCRIPTION

This module contains utilities for L<<with-sound>>.


=head1 DEPENDENCIES

Config::Simple (version 4.58 or later)

File::Path::Expand (version 1.02 or later)

File::Which (version 1.09 or later)

Test::Warn (version 0.24 or later)

Test::MockObject::Extends (version 1.20120301 or later)


=head1 AUTHOR

moznion  C<< <moznion@gmail.com> >>

Shinpei Maruyama C<< shinpeim[at]gmail.com> >>


=head1 CONTRIBUTOR

Syohei YOSHIDA C<< <syohex[at]gmail.com> >>


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2013, moznion C<< <moznion@gmail.com> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


Build.PL
use strict;
use warnings;
use Module::Build;
use Carp;

unless ( requires_external_bin('mpg123')
    || requires_external_bin('mpg321')
    || requires_external_bin('afplay') )
{
    print "! Unresolvable missing external dependency.\n";
    print "! Please install 'mpg123' or 'mpg321' seperately and try again.\n";
    print {*STDERR} "! NA: Unable to build distribution on this platform.\n";
    exit 0;
}

my $builder = Module::Build->new(
    name              => 'App-WithSound',
    module_name       => 'App::WithSound',
    script_files      => ['with-sound'],
    license           => 'perl',
    dist_author       => 'moznion <moznion@gmail.com>',
    dist_version_from => 'lib/App/WithSound.pm',
    build_requires    => {
        'Test::More'                => '0.98',
        'Test::MockObject::Extends' => '1.20120301',
        'Test::Warn'                => '0.24',
        'FindBin'                   => 0,
        'File::Spec::Functions'     => '3.40',
    },
    configure_requires => { 'Module::Build' => '0.40' },
    requires           => {
        perl                 => '5.008001',
        'File::Which'        => '0.05',
        'Config::Simple'     => '4.58',
        'File::Path::Expand' => '1.02',
        'IPC::Open3'         => 0,
        'Carp'               => 0,
    },
    no_index   => { 'directory' => ['inc'] },
    meta_merge => {
        'resources' => {
            'license'    => 'http://dev.perl.org/licenses/',
            'repository' => 'git://github.com/moznion/App--WithSound.git',
        },
    },
    test_files => ( -d '.git' || $ENV{RELEASE_TESTING} ) ? 't/ xt/' : 't/',
    recursive_test_files => 1,
    add_to_cleanup       => ['App-WithSound-*'],
    create_license       => 1,
);

$builder->create_build_script();

# Module::Install::Can
sub can_run {
    my $cmd = shift;
    require ExtUtils::MakeMaker;
    if ( $^O eq 'cygwin' ) {

        # MM->maybe_command is fixed in 6.51_01 for Cygwin.
        ExtUtils::MakeMaker->import(6.52);
    }

    my $_cmd = $cmd;
    return $_cmd if ( -x $_cmd or $_cmd = MM->maybe_command($_cmd) );

    for my $dir ( ( split /$Config::Config{path_sep}/x, $ENV{PATH} ), q{.} ) {
        next if $dir eq q{};
        my $abs = File::Spec->catfile( $dir, $cmd );
        return $abs if ( -x $abs or $abs = MM->maybe_command($abs) );
    }

    return;
} ## end sub can_run

# Module::Install::External
sub requires_external_bin {
    my ( $bin, $version ) = @_;
    if ($version) {
        croak 'requires_external_bin does not support versions yet';
    }

    # Locate the bin
    print "Locating required external dependency bin: $bin...";
    my $found_bin = can_run($bin);
    if ($found_bin) {
        print " found at $found_bin.\n";
        return 1;
    }
    print " missing.\n";
    return 0;
} ## end sub requires_external_bin
